File: python/78. Subsets.py
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []

        def convert2set(val, nums, length):
            set_list = []
            for i in range(length):
                if val & (1 << i):
                    set_list.append(nums[i])
            return set_list

        length = len(nums)
        count = 1 << length
        for i in range(count):
            res.append(convert2set(i, nums, length))
        return res

sol  = Solution()
print sol.subsets([1,2,3])

File: cpp/51-100/78. Subsets.cpp
class Solution {
public:
	vector<vector<int>> subsets(vector<int>& nums) {
		vector<vector<int>> res;
		vector<int> path;
		if (nums.empty())
			return res;
		dfs(0, 0, path, res, nums);
		return res;
	}
	void dfs(int depth, int k, vector<int> &path, vector<vector<int>> &res, vector<int> nums) {
			res.push_back(path);
		for (int i = k; i < nums.size(); i++)
		{
			path.push_back(nums[i]);
			dfs(depth, i + 1, path, res, nums);
			path.pop_back();
		}
	}
};